src/main/java/org/terasology/logic/inventory/StartingInventorySystem.java

```diff
@@ -99,58 +89,66 @@ private boolean validateItem(StartingInventoryComponent.InventoryItem item) {
     }
 
     private void addToInventory(EntityRef entityRef,
-                                StartingInventoryComponent.InventoryItem item,
-                                InventoryComponent inventoryComponent) {
-        String uri = item.uri;
-        int quantity = item.quantity;
-
-        final List<EntityRef> objects =
-                tryAsBlock(uri, quantity, item.items)
-                        .map(Optional::of)
-                        .orElseGet(() -> tryAsItem(uri, quantity))
-                        .orElse(Lists.newArrayList());
-
-        objects.forEach(o ->
-                inventoryManager.giveItem(entityRef, EntityRef.NULL, o)
-        );
+                                StartingInventoryComponent.InventoryItem item) {
+
+        Optional<Supplier<EntityRef>> possibleItem =
+                resolveAsBlock(item).map(Optional::of).orElseGet(() -> resolveAsItem(item));
```

Comment:
Is this understandable? Should we change this? Here are some alternatives:

```suggestion
        Optional<Supplier<EntityRef>> possibleItem = resolveAsBlock(item);
        if (!possibleItem.isPresent()) {
            possibleItem = resolveAsItem(item);
        }
```

```suggestion
        final Supplier<EntityRef> possibleItem = 
        		resolveAsBlock(item).orElse(resolveAsItem(item).orElse(null));
        if (possibleItem != null) {
            // ...
        }
```

Reply:
With `Optional::or` (Java 9) mentioned in the comments this part would become:

```java
final Supplier<EntityRef> possibleItem = 
    resolveAsBlock(item.uri).or(() -> resolveAsItem(item.uri))
```